src/backend/renderer/pixman/mod.rs

```diff
@@ -0,0 +1,1208 @@
+    ops::Deref,
+    os::unix::io::{AsFd, BorrowedFd},
+    rc::Rc,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use drm_fourcc::{DrmFormat, DrmFourcc, DrmModifier};
+use pixman::{Filter, FormatCode, Image, Operation, Repeat};
+use thiserror::Error;
+use tracing::warn;
+use wayland_server::{
+    protocol::{wl_buffer, wl_shm},
+    Weak,
+};
+
+use crate::{
+    backend::{
+        allocator::{dmabuf::Dmabuf, format::has_alpha, Buffer},
+        SwapBuffersError,
+    },
+    utils::{Buffer as BufferCoords, Physical, Rectangle, Scale, Size, Transform},
+    wayland::{compositor::SurfaceData, shm},
+};
+
+#[cfg(all(
+    feature = "wayland_frontend",
+    feature = "backend_egl",
+    feature = "use_system_lib"
+))]
+use super::ImportEgl;
+use super::{
+    sync::SyncPoint, Bind, DebugFlags, ExportMem, Frame, ImportDma, ImportDmaWl, ImportMem, ImportMemWl,
+    Offscreen, Renderer, Texture, TextureFilter, TextureMapping, Unbind,
+};
+
+bitflags::bitflags! {
+    struct DmaBufSyncFlags: std::ffi::c_ulonglong {
+        const READ = 1 << 0;
+        const WRITE = 2 << 0;
+        const START = 0 << 2;
+        const END = 1 << 2;
+    }
+}
+
+#[repr(C)]
+#[allow(non_camel_case_types)]
+struct dma_buf_sync {
+    flags: DmaBufSyncFlags,
+}
+
+struct DmabufIoctlSync {
+    sync: dma_buf_sync,
+}
+
+unsafe impl rustix::ioctl::Ioctl for DmabufIoctlSync {
+    type Output = ();
+
+    const OPCODE: rustix::ioctl::Opcode = rustix::ioctl::Opcode::write::<dma_buf_sync>(b'b', 0);
+
+    const IS_MUTATING: bool = true;
+
+    fn as_ptr(&mut self) -> *mut std::ffi::c_void {
+        &mut self.sync as *mut _ as *mut _
+    }
+
+    unsafe fn output_from_ptr(
+        _out: rustix::ioctl::IoctlOutput,
+        _extract_output: *mut std::ffi::c_void,
+    ) -> rustix::io::Result<Self::Output> {
+        Ok(())
+    }
+}
+
+fn dmabuf_sync_read_start(fd: impl AsFd) -> rustix::io::Result<()> {
+    let ioctl = DmabufIoctlSync {
+        sync: dma_buf_sync {
+            flags: DmaBufSyncFlags::READ | DmaBufSyncFlags::START,
+        },
+    };
+    unsafe { rustix::ioctl::ioctl(fd, ioctl) }
+}
+
+fn dmabuf_sync_read_end(fd: impl AsFd) -> rustix::io::Result<()> {
+    let ioctl = DmabufIoctlSync {
+        sync: dma_buf_sync {
+            flags: DmaBufSyncFlags::READ | DmaBufSyncFlags::END,
+        },
+    };
+    unsafe { rustix::ioctl::ioctl(fd, ioctl) }
+}
+
+fn dmabuf_sync_read_write_start(fd: impl AsFd) -> rustix::io::Result<()> {
+    let ioctl = DmabufIoctlSync {
+        sync: dma_buf_sync {
+            flags: DmaBufSyncFlags::READ | DmaBufSyncFlags::WRITE | DmaBufSyncFlags::START,
+        },
+    };
+    unsafe { rustix::ioctl::ioctl(fd, ioctl) }
+}
+
+fn dmabuf_sync_read_write_end(fd: impl AsFd) -> rustix::io::Result<()> {
+    let ioctl = DmabufIoctlSync {
+        sync: dma_buf_sync {
+            flags: DmaBufSyncFlags::READ | DmaBufSyncFlags::WRITE | DmaBufSyncFlags::END,
+        },
+    };
+    unsafe { rustix::ioctl::ioctl(fd, ioctl) }
+}
```

Comment:
Can we put this somewhere more accessible and perhaps even make it public?

Reply:
sure, as this is related to dmabuf we could put it there I guess?

Reply:
sounds good.